File: go/pkg/chatgear/stats.go
package chatgear

import (
	"bytes"
	"encoding/json"
	"slices"

	"github.com/haivivi/giztoy/go/pkg/encoding"
	"github.com/haivivi/giztoy/go/pkg/jsontime"
)

// StatsChanges represents changes to device statistics.
type StatsChanges struct {
	Time jsontime.Milli

	LastResetAt   *jsontime.Milli
	Battery       *Battery
	SystemVersion *SystemVersion
	Volume        *Volume
	Brightness    *Brightness
	LightMode     *LightMode
	Cellular      *ConnectedCellular
	WifiNetwork   *ConnectedWifi
	WifiStore     *StoredWifiList
	ReadNFCTag    *ReadNFCTag
	PairStatus    *PairStatus
	Shaking       *Shaking
}

// StatsEvent converts changes to a full stats event.
func (c *StatsChanges) StatsEvent() *StatsEvent {
	if c == nil {
		return nil
	}
	e := &StatsEvent{
		Time:          c.Time,
		Battery:       c.Battery,
		SystemVersion: c.SystemVersion,
		Volume:        c.Volume,
		Brightness:    c.Brightness,
		LightMode:     c.LightMode,
		WifiNetwork:   c.WifiNetwork,
		WifiStore:     c.WifiStore,
		ReadNFCTag:    c.ReadNFCTag,
		PairStatus:    c.PairStatus,
		Shaking:       c.Shaking,
	}
	if c.LastResetAt != nil {
		e.LastResetAt = *c.LastResetAt
	}
	return e
}

// StatsEvent contains device statistics.
type StatsEvent struct {
	Time          jsontime.Milli     `json:"time"`
	LastResetAt   jsontime.Milli     `json:"last_reset_at,omitzero"`
	Battery       *Battery           `json:"battery,omitzero"`
	SystemVersion *SystemVersion     `json:"system_version,omitzero"`
	Volume        *Volume            `json:"volume,omitzero"`
	Brightness    *Brightness        `json:"brightness,omitzero"`
	LightMode     *LightMode         `json:"light_mode,omitzero"`
	Cellular      *ConnectedCellular `json:"cellular_network,omitzero"`
	WifiNetwork   *ConnectedWifi     `json:"wifi_network,omitzero"`
	WifiStore     *StoredWifiList    `json:"wifi_store,omitzero"`
	ReadNFCTag    *ReadNFCTag        `json:"read_nfc_tag,omitzero"`
	PairStatus    *PairStatus        `json:"pair_status,omitzero"`
	Shaking       *Shaking           `json:"shaking,omitzero"`
}

// SystemVersion contains system version information.
type SystemVersion struct {
	CurrentVersion       string             `json:"current_version,omitzero"`
	InstallingPercentage float64            `json:"installing_percentage,omitzero"`
	InstallingVersion    string             `json:"installing_version,omitzero"`
	Components           []ComponentVersion `json:"components,omitzero"`
	UpdateAt             jsontime.Milli     `json:"update_at,omitzero"`
}

// ComponentVersion contains version info for a system component.
type ComponentVersion struct {
	Name           string `json:"name"`
	CurrentVersion string `json:"current_version,omitzero"`
}

// Brightness contains display brightness settings.
type Brightness struct {
	Percentage float64        `json:"percentage"`
	UpdateAt   jsontime.Milli `json:"update_at"`
}

// LightMode contains light mode settings.
type LightMode struct {
	Mode     string         `json:"mode"`
	UpdateAt jsontime.Milli `json:"update_at"`
}

// Volume contains audio volume settings.
type Volume struct {
	Percentage float64        `json:"percentage"`
	UpdateAt   jsontime.Milli `json:"update_at"`
}

// Ping contains network ping statistics.
type Ping struct {
	RecvAt    jsontime.Milli `json:"recv_at"`
	SendAt    jsontime.Milli `json:"send_at"`
	Timestamp jsontime.Milli `json:"timestamp"`
	Latency   float64        `json:"latency"`
}

// ConnectedCellular contains cellular network information.
type ConnectedCellular struct {
	DNS     []string `json:"dns,omitzero"`
	Gateway string   `json:"gateway,omitzero"`
	IP      string   `json:"ip,omitzero"`
	NetMask string   `json:"mask,omitzero"`
	RSSI    float64  `json:"rssi,omitzero"`
	Ping    *Ping    `json:"ping,omitzero"`
}

// ConnectedWifi contains WiFi network information.
type ConnectedWifi struct {
	BSSID    string   `json:"bssid,omitzero"`
	DNS      []string `json:"dns,omitzero"`
	Gateway  string   `json:"gateway,omitzero"`
	IP       string   `json:"ip,omitzero"`
	NetMask  string   `json:"mask,omitzero"`
	Mac      string   `json:"mac,omitzero"`
	RSSI     float64  `json:"rssi,omitzero"`
	Security string   `json:"security,omitzero"`
	SSID     string   `json:"ssid,omitzero"`
	Ping     *Ping    `json:"ping,omitzero"`
}

// WifiStoreItem contains information about a stored WiFi network.
type WifiStoreItem struct {
	LastConnectAt jsontime.Milli `json:"last_connect_at"`
	Security      string         `json:"security,omitzero"`
	SSID          string         `json:"ssid,omitzero"`
}

// StoredWifiList contains a list of stored WiFi networks.
type StoredWifiList struct {
	List     []WifiStoreItem `json:"list,omitzero"`
	UpdateAt jsontime.Milli  `json:"update_at"`
}

// Battery contains battery status information.
type Battery struct {
	CurrentCapacity  float64 `json:"current_capacity,omitzero"`
	CycleCount       float64 `json:"cycle_count,omitzero"`
	IsCharging       bool    `json:"is_charging,omitzero"`
	OriginalCapacity float64 `json:"original_capacity,omitzero"`
	Percentage       float64 `json:"percentage,omitzero"`
	Temperature      float64 `json:"temperature,omitzero"`
	Voltage          float64 `json:"voltage,omitzero"`
}

// ReadNFCTag contains NFC tag read data.
type ReadNFCTag struct {
	Tags []*NFCTag `json:"tags,omitzero"`
}

// NFCTag contains information about an NFC tag.
type NFCTag struct {
	UID        string                `json:"uid,omitzero"`
	Type       string                `json:"type,omitzero"`
	RawData    encoding.StdBase64Data `json:"raw_data,omitzero"`
	DataFormat string                `json:"data_format,omitzero"`
	RSSI       float32               `json:"rssi,omitzero"`
	UpdateAt   jsontime.Milli        `json:"update_at,omitzero"`
}

// PairStatus contains device pairing status.
type PairStatus struct {
	PairWith string         `json:"pair_with,omitzero"`
	UpdateAt jsontime.Milli `json:"update_at,omitzero"`
}

// Shaking contains device shaking detection data.
type Shaking struct {
	Level float64 `json:"level"`
}

// Clone returns a deep copy of the stats event.
func (e *StatsEvent) Clone() *StatsEvent {
	if e == nil {
		return nil
	}
	v := *e
	v.Battery = clonePtr(e.Battery)
	v.SystemVersion = e.SystemVersion.clone()
	v.Volume = clonePtr(e.Volume)
	v.Brightness = clonePtr(e.Brightness)
	v.LightMode = clonePtr(e.LightMode)
	v.WifiNetwork = e.WifiNetwork.clone()
	v.WifiStore = e.WifiStore.clone()
	v.ReadNFCTag = e.ReadNFCTag.clone()
	v.PairStatus = clonePtr(e.PairStatus)
	v.Shaking = clonePtr(e.Shaking)
	return &v
}

func (sv *SystemVersion) clone() *SystemVersion {
	if sv == nil {
		return nil
	}
	v := *sv
	v.Components = slices.Clone(sv.Components)
	return &v
}

func (cc *ConnectedCellular) clone() *ConnectedCellular {
	if cc == nil {
		return nil
	}
	v := *cc
	v.DNS = slices.Clone(cc.DNS)
	v.Ping = clonePtr(cc.Ping)
	return &v
}

func (cw *ConnectedWifi) clone() *ConnectedWifi {
	if cw == nil {
		return nil
	}
	v := *cw
	v.DNS = slices.Clone(cw.DNS)
	v.Ping = clonePtr(cw.Ping)
	return &v
}

func (swl *StoredWifiList) clone() *StoredWifiList {
	if swl == nil {
		return nil
	}
	v := *swl
	v.List = slices.Clone(swl.List)
	return &v
}

func (rnt *ReadNFCTag) clone() *ReadNFCTag {
	if rnt == nil {
		return nil
	}
	v := &ReadNFCTag{
		Tags: make([]*NFCTag, 0, len(rnt.Tags)),
	}
	for _, tag := range rnt.Tags {
		v.Tags = append(v.Tags, tag.clone())
	}
	return v
}

func (nt *NFCTag) clone() *NFCTag {
	if nt == nil {
		return nil
	}
	return &NFCTag{
		UID:        nt.UID,
		Type:       nt.Type,
		RawData:    bytes.Clone(nt.RawData),
		DataFormat: nt.DataFormat,
		RSSI:       nt.RSSI,
		UpdateAt:   nt.UpdateAt,
	}
}

// MarshalJSON implements json.Marshaler.
func (rnt *ReadNFCTag) MarshalJSON() ([]byte, error) {
	return json.Marshal(rnt.Tags)
}

// UnmarshalJSON implements json.Unmarshaler.
func (rnt *ReadNFCTag) UnmarshalJSON(data []byte) error {
	*rnt = ReadNFCTag{}
	return json.Unmarshal(data, &rnt.Tags)
}

// MergeWith merges another stats event into this one.
// Returns the changes if any fields were updated.
func (e *StatsEvent) MergeWith(other *StatsEvent) *StatsChanges {
	if other.Time.Before(e.Time) {
		return nil
	}
	e.Time = other.Time

	var diff StatsChanges

	if other.LastResetAt.After(e.LastResetAt) {
		e.LastResetAt = other.LastResetAt
		v := other.LastResetAt
		diff.LastResetAt = &v
	}

	switch {
	case other.SystemVersion == nil:
	case e.SystemVersion == nil,
		!other.SystemVersion.UpdateAt.Before(e.SystemVersion.UpdateAt):
		e.SystemVersion = other.SystemVersion
		diff.SystemVersion = other.SystemVersion.clone()
	}

	switch {
	case other.Volume == nil:
	case e.Volume == nil,
		!other.Volume.UpdateAt.Before(e.Volume.UpdateAt):
		e.Volume = other.Volume
		diff.Volume = clonePtr(other.Volume)
	}

	switch {
	case other.Brightness == nil:
	case e.Brightness == nil,
		!other.Brightness.UpdateAt.Before(e.Brightness.UpdateAt):
		e.Brightness = other.Brightness
		diff.Brightness = clonePtr(other.Brightness)
	}

	switch {
	case other.LightMode == nil:
	case e.LightMode == nil,
		!other.LightMode.UpdateAt.Before(e.LightMode.UpdateAt):
		e.LightMode = other.LightMode
		diff.LightMode = clonePtr(other.LightMode)
	}

	switch {
	case other.PairStatus == nil:
	case e.PairStatus == nil,
		!other.PairStatus.UpdateAt.Before(e.PairStatus.UpdateAt):
		e.PairStatus = other.PairStatus
		diff.PairStatus = clonePtr(other.PairStatus)
	}

	switch {
	case other.Battery == nil:
	case !other.Battery.Equal(e.Battery):
		e.Battery = other.Battery
		diff.Battery = clonePtr(other.Battery)
	}

	switch {
	case other.WifiStore == nil:
	case e.WifiStore == nil,
		!other.WifiStore.UpdateAt.Before(e.WifiStore.UpdateAt):
		e.WifiStore = other.WifiStore
		diff.WifiStore = other.WifiStore.clone()
	}

	switch {
	case other.Cellular == nil:
	case !e.Cellular.Equal(other.Cellular):
		e.Cellular = other.Cellular
		diff.Cellular = other.Cellular.clone()
	}

	switch {
	case other.WifiNetwork == nil:
	case !e.WifiNetwork.Equal(other.WifiNetwork):
		e.WifiNetwork = other.WifiNetwork
		diff.WifiNetwork = other.WifiNetwork.clone()
	}

	switch {
	case other.ReadNFCTag == nil:
	case !other.ReadNFCTag.Equal(e.ReadNFCTag):
		e.ReadNFCTag = other.ReadNFCTag
		diff.ReadNFCTag = other.ReadNFCTag.clone()
	}

	switch {
	case other.Shaking == nil:
	case e.Shaking == nil,
		!other.Shaking.Equal(e.Shaking):
		e.Shaking = other.Shaking
		diff.Shaking = clonePtr(other.Shaking)
	}

	if diff == (StatsChanges{}) {
		return nil
	}
	diff.Time = e.Time
	return &diff
}

// Equal methods for comparison

func (s *Shaking) Equal(other *Shaking) bool {
	if s == nil || other == nil {
		return s == other
	}
	return s.Level == other.Level
}

func (b *Battery) Equal(other *Battery) bool {
	if b == nil || other == nil {
		return b == other
	}
	return *b == *other
}

func (p *Ping) Equal(other *Ping) bool {
	if p == nil || other == nil {
		return p == other
	}
	return *p == *other
}

func (cc *ConnectedCellular) Equal(other *ConnectedCellular) bool {
	if cc == nil || other == nil {
		return cc == other
	}
	ok := cc.Gateway == other.Gateway &&
		cc.IP == other.IP &&
		cc.NetMask == other.NetMask &&
		cc.RSSI == other.RSSI &&
		cc.Ping.Equal(other.Ping)
	if !ok {
		return false
	}
	if len(cc.DNS) != len(other.DNS) {
		return false
	}
	for i, v := range cc.DNS {
		if v != other.DNS[i] {
			return false
		}
	}
	return true
}

func (cw *ConnectedWifi) Equal(other *ConnectedWifi) bool {
	if cw == nil || other == nil {
		return cw == other
	}
	ok := cw.BSSID == other.BSSID &&
		cw.Gateway == other.Gateway &&
		cw.IP == other.IP &&
		cw.Mac == other.Mac &&
		cw.NetMask == other.NetMask &&
		cw.RSSI == other.RSSI &&
		cw.Security == other.Security &&
		cw.SSID == other.SSID &&
		cw.Ping.Equal(other.Ping)
	if !ok {
		return false
	}
	if len(cw.DNS) != len(other.DNS) {
		return false
	}
	for i, v := range cw.DNS {
		if v != other.DNS[i] {
			return false
		}
	}
	return true
}

func (nt *NFCTag) Equal(other *NFCTag) bool {
	if nt == nil || other == nil {
		return nt == other
	}
	return nt.UID == other.UID
}

func (rnt *ReadNFCTag) Equal(other *ReadNFCTag) bool {
	if rnt == nil || other == nil {
		return rnt == other
	}

	rntUids := make(map[string]struct{})
	for _, tag := range rnt.Tags {
		rntUids[tag.UID] = struct{}{}
	}

	otherUids := make(map[string]struct{})
	for _, tag := range other.Tags {
		otherUids[tag.UID] = struct{}{}
	}

	if len(rntUids) != len(otherUids) {
		return false
	}

	for uid := range rntUids {
		if _, ok := otherUids[uid]; !ok {
			return false
		}
	}

	return true
}

// clonePtr creates a shallow copy of a pointer value.
func clonePtr[T any](v *T) *T {
	if v == nil {
		return nil
	}
	v2 := *v
	return &v2
}

File: rust/chatgear/src/stats.rs
//! Device statistics types.

use giztoy_encoding::StdBase64Data;
use giztoy_jsontime::Milli;
use serde::{Deserialize, Serialize};

/// Contains device statistics.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct GearStatsEvent {
    /// Event timestamp.
    pub time: Milli,

    /// Last device reset time.
    #[serde(default, skip_serializing_if = "Milli::is_zero")]
    pub last_reset_at: Milli,

    /// Battery status.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub battery: Option<Battery>,

    /// System version information.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub system_version: Option<SystemVersion>,

    /// Audio volume settings.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub volume: Option<Volume>,

    /// Display brightness settings.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub brightness: Option<Brightness>,

    /// Light mode settings.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub light_mode: Option<LightMode>,

    /// Cellular network information.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cellular_network: Option<ConnectedCellular>,

    /// WiFi network information.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub wifi_network: Option<ConnectedWifi>,

    /// Stored WiFi list.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub wifi_store: Option<StoredWifiList>,

    /// NFC tag read data.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub read_nfc_tag: Option<ReadNFCTag>,

    /// Device pairing status.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pair_status: Option<PairStatus>,

    /// Device shaking detection data.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub shaking: Option<Shaking>,
}

impl GearStatsEvent {
    /// Creates a new empty stats event.
    pub fn new() -> Self {
        Self {
            time: Milli::now(),
            ..Default::default()
        }
    }

    /// Returns a deep copy of the stats event.
    pub fn clone_deep(&self) -> Self {
        self.clone()
    }

    /// Merges another stats event into this one.
    /// Returns the changes if any fields were updated.
    pub fn merge_with(&mut self, other: &GearStatsEvent) -> Option<GearStatsChanges> {
        if other.time.before(&self.time) {
            return None;
        }
        self.time = other.time;

        let mut diff = GearStatsChanges::default();
        let mut has_changes = false;

        // Last reset at
        if other.last_reset_at.after(&self.last_reset_at) {
            self.last_reset_at = other.last_reset_at;
            diff.last_reset_at = Some(other.last_reset_at);
            has_changes = true;
        }

        // System version
        if let Some(ref other_sv) = other.system_version {
            if self.system_version.is_none()
                || !other_sv.update_at.before(&self.system_version.as_ref().unwrap().update_at)
            {
                self.system_version = Some(other_sv.clone());
                diff.system_version = Some(other_sv.clone());
                has_changes = true;
            }
        }

        // Volume
        if let Some(ref other_vol) = other.volume {
            if self.volume.is_none()
                || !other_vol.update_at.before(&self.volume.as_ref().unwrap().update_at)
            {
                self.volume = Some(other_vol.clone());
                diff.volume = Some(other_vol.clone());
                has_changes = true;
            }
        }

        // Brightness
        if let Some(ref other_br) = other.brightness {
            if self.brightness.is_none()
                || !other_br.update_at.before(&self.brightness.as_ref().unwrap().update_at)
            {
                self.brightness = Some(other_br.clone());
                diff.brightness = Some(other_br.clone());
                has_changes = true;
            }
        }

        // Light mode
        if let Some(ref other_lm) = other.light_mode {
            if self.light_mode.is_none()
                || !other_lm.update_at.before(&self.light_mode.as_ref().unwrap().update_at)
            {
                self.light_mode = Some(other_lm.clone());
                diff.light_mode = Some(other_lm.clone());
                has_changes = true;
            }
        }

        // Pair status
        if let Some(ref other_ps) = other.pair_status {
            if self.pair_status.is_none()
                || !other_ps.update_at.before(&self.pair_status.as_ref().unwrap().update_at)
            {
                self.pair_status = Some(other_ps.clone());
                diff.pair_status = Some(other_ps.clone());
                has_changes = true;
            }
        }

        // Battery
        if let Some(ref other_bat) = other.battery {
            if self.battery.as_ref() != Some(other_bat) {
                self.battery = Some(other_bat.clone());
                diff.battery = Some(other_bat.clone());
                has_changes = true;
            }
        }

        // WiFi store
        if let Some(ref other_ws) = other.wifi_store {
            if self.wifi_store.is_none()
                || !other_ws.update_at.before(&self.wifi_store.as_ref().unwrap().update_at)
            {
                self.wifi_store = Some(other_ws.clone());
                diff.wifi_store = Some(other_ws.clone());
                has_changes = true;
            }
        }

        // Cellular
        if let Some(ref other_cell) = other.cellular_network {
            if self.cellular_network.as_ref() != Some(other_cell) {
                self.cellular_network = Some(other_cell.clone());
                diff.cellular = Some(other_cell.clone());
                has_changes = true;
            }
        }

        // WiFi network
        if let Some(ref other_wifi) = other.wifi_network {
            if self.wifi_network.as_ref() != Some(other_wifi) {
                self.wifi_network = Some(other_wifi.clone());
                diff.wifi_network = Some(other_wifi.clone());
                has_changes = true;
            }
        }

        // NFC tag
        if let Some(ref other_nfc) = other.read_nfc_tag {
            if self.read_nfc_tag.as_ref() != Some(other_nfc) {
                self.read_nfc_tag = Some(other_nfc.clone());
                diff.read_nfc_tag = Some(other_nfc.clone());
                has_changes = true;
            }
        }

        // Shaking
        if let Some(ref other_shake) = other.shaking {
            if self.shaking.as_ref() != Some(other_shake) {
                self.shaking = Some(other_shake.clone());
                diff.shaking = Some(other_shake.clone());
                has_changes = true;
            }
        }

        if has_changes {
            diff.time = self.time;
            Some(diff)
        } else {
            None
        }
    }
}

/// Represents changes to device statistics.
#[derive(Debug, Clone, Default)]
pub struct GearStatsChanges {
    pub time: Milli,
    pub last_reset_at: Option<Milli>,
    pub battery: Option<Battery>,
    pub system_version: Option<SystemVersion>,
    pub volume: Option<Volume>,
    pub brightness: Option<Brightness>,
    pub light_mode: Option<LightMode>,
    pub cellular: Option<ConnectedCellular>,
    pub wifi_network: Option<ConnectedWifi>,
    pub wifi_store: Option<StoredWifiList>,
    pub read_nfc_tag: Option<ReadNFCTag>,
    pub pair_status: Option<PairStatus>,
    pub shaking: Option<Shaking>,
}

impl GearStatsChanges {
    /// Converts changes to a full stats event.
    pub fn to_event(&self) -> GearStatsEvent {
        GearStatsEvent {
            time: self.time,
            last_reset_at: self.last_reset_at.unwrap_or_default(),
            battery: self.battery.clone(),
            system_version: self.system_version.clone(),
            volume: self.volume.clone(),
            brightness: self.brightness.clone(),
            light_mode: self.light_mode.clone(),
            cellular_network: self.cellular.clone(),
            wifi_network: self.wifi_network.clone(),
            wifi_store: self.wifi_store.clone(),
            read_nfc_tag: self.read_nfc_tag.clone(),
            pair_status: self.pair_status.clone(),
            shaking: self.shaking.clone(),
        }
    }
}

// ============================================================================
// Statistics Types
// ============================================================================

/// Contains system version information.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct SystemVersion {
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub current_version: String,
    #[serde(default, skip_serializing_if = "is_zero_f64")]
    pub installing_percentage: f64,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub installing_version: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub components: Vec<ComponentVersion>,
    #[serde(default, skip_serializing_if = "Milli::is_zero")]
    pub update_at: Milli,
}

fn is_zero_f64(v: &f64) -> bool {
    *v == 0.0
}

/// Contains version info for a system component.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ComponentVersion {
    pub name: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub current_version: String,
}

/// Contains display brightness settings.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct Brightness {
    pub percentage: f64,
    pub update_at: Milli,
}

/// Contains light mode settings.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct LightMode {
    pub mode: String,
    pub update_at: Milli,
}

/// Contains audio volume settings.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct Volume {
    pub percentage: f64,
    pub update_at: Milli,
}

/// Contains network ping statistics.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct Ping {
    pub recv_at: Milli,
    pub send_at: Milli,
    pub timestamp: Milli,
    pub latency: f64,
}

/// Contains cellular network information.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct ConnectedCellular {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub dns: Vec<String>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub gateway: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub ip: String,
    #[serde(default, rename = "mask", skip_serializing_if = "String::is_empty")]
    pub net_mask: String,
    #[serde(default, skip_serializing_if = "is_zero_f64")]
    pub rssi: f64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ping: Option<Ping>,
}

/// Contains WiFi network information.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct ConnectedWifi {
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub bssid: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub dns: Vec<String>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub gateway: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub ip: String,
    #[serde(default, rename = "mask", skip_serializing_if = "String::is_empty")]
    pub net_mask: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub mac: String,
    #[serde(default, skip_serializing_if = "is_zero_f64")]
    pub rssi: f64,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub security: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub ssid: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ping: Option<Ping>,
}

/// Contains information about a stored WiFi network.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct WifiStoreItem {
    pub last_connect_at: Milli,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub security: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub ssid: String,
}

/// Contains a list of stored WiFi networks.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct StoredWifiList {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub list: Vec<WifiStoreItem>,
    pub update_at: Milli,
}

/// Contains battery status information.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct Battery {
    #[serde(default, skip_serializing_if = "is_zero_f64")]
    pub current_capacity: f64,
    #[serde(default, skip_serializing_if = "is_zero_f64")]
    pub cycle_count: f64,
    #[serde(default)]
    pub is_charging: bool,
    #[serde(default, skip_serializing_if = "is_zero_f64")]
    pub original_capacity: f64,
    #[serde(default, skip_serializing_if = "is_zero_f64")]
    pub percentage: f64,
    #[serde(default, skip_serializing_if = "is_zero_f64")]
    pub temperature: f64,
    #[serde(default, skip_serializing_if = "is_zero_f64")]
    pub voltage: f64,
}

/// Contains NFC tag read data.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(transparent)]
pub struct ReadNFCTag {
    pub tags: Vec<NFCTag>,
}

/// Contains information about an NFC tag.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct NFCTag {
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub uid: String,
    #[serde(default, rename = "type", skip_serializing_if = "String::is_empty")]
    pub tag_type: String,
    #[serde(default, skip_serializing_if = "StdBase64Data::is_empty")]
    pub raw_data: StdBase64Data,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub data_format: String,
    #[serde(default, skip_serializing_if = "is_zero_f32")]
    pub rssi: f32,
    #[serde(default, skip_serializing_if = "Milli::is_zero")]
    pub update_at: Milli,
}

fn is_zero_f32(v: &f32) -> bool {
    *v == 0.0
}

/// Contains device pairing status.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct PairStatus {
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub pair_with: String,
    #[serde(default, skip_serializing_if = "Milli::is_zero")]
    pub update_at: Milli,
}

/// Contains device shaking detection data.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct Shaking {
    pub level: f64,
}

#[cfg(test)]
mod stats_tests {
    use super::*;

    #[test]
    fn test_gear_stats_event_new() {
        let event = GearStatsEvent::new();
        assert!(!event.time.is_zero());
    }

    #[test]
    fn test_gear_stats_event_serialize() {
        let mut event = GearStatsEvent::new();
        event.battery = Some(Battery {
            percentage: 75.0,
            is_charging: true,
            ..Default::default()
        });

        let json = serde_json::to_string(&event).unwrap();
        assert!(json.contains("battery"));
        assert!(json.contains("75"));

        let restored: GearStatsEvent = serde_json::from_str(&json).unwrap();
        assert_eq!(restored.battery.as_ref().unwrap().percentage, 75.0);
    }

    #[test]
    fn test_volume_serialize() {
        let vol = Volume {
            percentage: 50.0,
            update_at: Milli::now(),
        };

        let json = serde_json::to_string(&vol).unwrap();
        let restored: Volume = serde_json::from_str(&json).unwrap();
        assert_eq!(restored.percentage, 50.0);
    }

    #[test]
    fn test_wifi_serialize() {
        let wifi = ConnectedWifi {
            ssid: "TestNetwork".to_string(),
            rssi: -45.0,
            ip: "192.168.1.100".to_string(),
            ..Default::default()
        };

        let json = serde_json::to_string(&wifi).unwrap();
        assert!(json.contains("TestNetwork"));
        assert!(json.contains("-45"));

        let restored: ConnectedWifi = serde_json::from_str(&json).unwrap();
        assert_eq!(restored.ssid, "TestNetwork");
    }

    #[test]
    fn test_nfc_tag_serialize() {
        let tag = NFCTag {
            uid: "04:AB:CD:EF".to_string(),
            tag_type: "NTAG215".to_string(),
            raw_data: StdBase64Data::from(vec![0x01, 0x02, 0x03]),
            ..Default::default()
        };

        let json = serde_json::to_string(&tag).unwrap();
        assert!(json.contains("04:AB:CD:EF"));

        let restored: NFCTag = serde_json::from_str(&json).unwrap();
        assert_eq!(restored.uid, "04:AB:CD:EF");
    }

    #[test]
    fn test_stats_merge() {
        let mut event1 = GearStatsEvent::new();
        event1.battery = Some(Battery {
            percentage: 50.0,
            ..Default::default()
        });

        std::thread::sleep(std::time::Duration::from_millis(1));

        let mut event2 = GearStatsEvent::new();
        event2.battery = Some(Battery {
            percentage: 75.0,
            ..Default::default()
        });

        let changes = event1.merge_with(&event2);
        assert!(changes.is_some());
        assert_eq!(event1.battery.as_ref().unwrap().percentage, 75.0);
    }
}
